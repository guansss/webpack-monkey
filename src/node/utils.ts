// @ts-ignore
// prettier-ignore
import { getDescriptionFile, getRequiredVersionFromDescriptionFile } from "webpack/lib/sharing/utils";

import { isArray, isBoolean, isNil, isObject, isString } from "lodash"
import { Compilation, sources } from "webpack"
import { getGMAPIs } from "../shared/GM"
import { META_FIELDS, META_FIELDS_WITH_LOCALIZATION, UserscriptMeta } from "../shared/meta"
import { includes } from "../shared/utils"

export function getPackageDepVersion(packageJson: any, dep: string): string | undefined {
  if (!isObject(packageJson)) {
    return undefined
  }

  return getRequiredVersionFromDescriptionFile(packageJson, dep)
}

export function getPackageJson(fs: Compilation["inputFileSystem"], context: string) {
  return new Promise<{ data: object; path: string }>((resolve, reject) => {
    getDescriptionFile(fs, context, ["package.json"], (err: any, result: any) => {
      if (err) {
        reject(err)
      } else {
        resolve(result)
      }
    })
  })
}

export function traverseAndFindSource(
  source: unknown,
  cb: (source: sources.Source) => void | true
): sources.Source | undefined {
  if (!source || !(source instanceof sources.Source)) {
    return undefined
  }

  const found = cb(source)

  if (found === true) {
    return source
  }

  if ((source as any)._children) {
    for (const child of (source as any)._children as unknown[]) {
      const result = traverseAndFindSource(child, cb)

      if (result) {
        return result
      }
    }
  }

  return traverseAndFindSource((source as any)._source, cb)
}

export function generateMetaBlock(
  source: string,
  meta: UserscriptMeta,
  { requires }: { requires?: string[] } = {}
) {
  let metaBlock = "// ==UserScript==\n"
  const fieldPrefix = "// @"

  const metaFields = Object.keys(meta)
  const maxFieldLength = Math.max(
    ...["grant", "require", ...metaFields].map((field) => field.length)
  )
  const indentSize = fieldPrefix.length + maxFieldLength + 2
  const indentEnd = (str: string) => str.padEnd(indentSize, " ")

  function putField(field: string, value: string | string[] | boolean) {
    let line = fieldPrefix + field

    if (isString(value)) {
      line = indentEnd(line) + value
    } else if (isArray(value)) {
      if (!value.length) {
        return
      }

      line = value.map((v) => indentEnd(line) + v).join("\n")
    } else if (isBoolean(value)) {
      // ignore false value
      if (!value) {
        return
      }
    } else {
      console.warn("Unknown type of value:", value)
    }

    metaBlock += line + "\n"
  }

  for (const field of META_FIELDS) {
    if (includes(META_FIELDS_WITH_LOCALIZATION, field)) {
      const value = meta[field]

      if (value) {
        if (isString(value)) {
          putField(field, value)
        } else {
          for (const lang in value) {
            putField(field + (lang === "default" ? "" : ":" + lang), value[lang]!)
          }
        }
      }
    } else if (field === "grant") {
      putField(
        "grant",
        getGMAPIs().filter((api) => source.includes(api))
      )

      if (meta.grant) {
        putField("grant", meta.grant)
      }
    } else if (field === "require") {
      if (requires) {
        putField("require", requires)
      }
    } else {
      if (!isNil(meta[field])) {
        putField(field, meta[field]!)
      }
    }
  }

  metaBlock += "// ==/UserScript=="

  return metaBlock
}

export function getUnnamedUrlExternalErrorMessage(url?: string) {
  const exampleUrl = url || "https://example.com"
  return `Unexpected reference to unnamed external module with URL "${url || "<unknown URL>"}".
This happens when you import a module from a URL
but do not specify an identifier for it,
e.g. "import foo from "${exampleUrl}", which will cause
runtime errors in the generated code. To fix this,
either add a universally unique identifier to the import statement,
e.g. "import foo from "foo@${exampleUrl}",
or do not import anything from it,
e.g. "import "${exampleUrl}".`.replace(/\n/g, " ")
}
