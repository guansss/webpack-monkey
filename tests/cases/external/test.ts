import { it } from "@jest/globals"
import path from "path"
import { monkeyWebpack } from "../../../src"
import { testBuild, withCommonConfig } from "../../utils/webpack"

const config = withCommonConfig({
  entry: path.resolve(__dirname, "index.js"),
  output: {
    path: path.resolve(__dirname, "dist"),
  },
  externals: {
    "ex-named": "exNamed",
    ex: "EX",
  }
})

it("build", () => testBuild(monkeyWebpack()(config)), 100000000)
