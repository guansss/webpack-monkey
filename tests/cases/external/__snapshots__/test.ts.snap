// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build 1`] = `
"// ==UserScript==
// @name     External test
// @require  https://unpkg.com/ex-namedundefined
// @require  https://ex-url-named
// @require  https://ex-url
// @require  https://ex-url-global-named
// @require  https://ex-url-global
// @require  https://ex-url-global2
// @require  https://ex-url-global3
// @require  https://unpkg.com/exundefined
// @match    *://*/*
// @version  1.0.0
// ==/UserScript==

;(() => {
  "use strict"
  // The require scope
  var __webpack_require__ = {}
  /* webpack/runtime/compat get default export */
  ;(() => {
    // getDefaultExport function for compatibility with non-harmony modules
    __webpack_require__.n = (module) => {
      var getter = module && module.__esModule ? () => module["default"] : () => module
      __webpack_require__.d(getter, { a: getter })
      return getter
    }
  })()
  /* webpack/runtime/define property getters */
  ;(() => {
    // define getter functions for harmony exports
    __webpack_require__.d = (exports, definition) => {
      for (var key in definition)
        __webpack_require__.o(definition, key) &&
          !__webpack_require__.o(exports, key) &&
          Object.defineProperty(exports, key, { enumerable: true, get: definition[key] })
    }
  })()
  /* webpack/runtime/hasOwnProperty shorthand */
  ;(() => {
    __webpack_require__.o = (obj, prop) => Object.prototype.hasOwnProperty.call(obj, prop)
  })()

  const external_exNamed_namespaceObject = exNamed
  var external_exNamed_default = __webpack_require__.n(external_exNamed_namespaceObject)

  const external_exUrlNamed_namespaceObject = exUrlNamed
  var external_exUrlNamed_default = __webpack_require__.n(external_exUrlNamed_namespaceObject)

  const ex_url_namespaceObject = void 0
  var ex_url_default = __webpack_require__.n(ex_url_namespaceObject)

  exUrlGlobalNamed

  EX

  window.externalTestResult = [
    [external_exNamed_default().x1, external_exNamed_namespaceObject.x1],
    [ex_url_default().x2, ex_url_namespaceObject.x2],
    [external_exUrlNamed_default().x3, external_exUrlNamed_namespaceObject.x3],
    [exUrlGlobal.x1],
    [exUrlGlobalNamed.x1],
  ]
})()
"
`;
